mne/beamformer/_dics.py

```diff
@@ -421,6 +423,97 @@ def apply_dics_epochs(epochs, filters, return_generator=False, verbose=None):
     return stcs
 
 
+@verbose
+def apply_dics_tfr(epochs, filters, freqs, method='morlet', average=False,
+                   n_csd=256, n_overlap=0, verbose=None, **csd_kwargs):
+    """Apply Dynamic Imaging of Coherent Sources (DICS) beamformer weights.
+
+    Apply Dynamic Imaging of Coherent Sources (DICS) beamformer weights
+    on single trial time-frequency data.
+
+    Parameters
+    ----------
+    epochs : Epochs
+        Single trial epochs.
+    filters : instance of Beamformer
+        DICS spatial filter (beamformer weights)
+        Filter weights returned from :func:`make_dics`.
+    freqs : ndarray, shape (n_freqs,)
+        The frequencies over which to compute cross-spectral density.
+    method : 'morlet' | 'fourier' | 'multitaper'
+        The cross-spectral density method to use.
+    average : bool
+        Whether to average across epochs. Default False.
+    n_csd : int
+        The number of samples to use for the cross-spectral density
+        windows.
+    n_overlap : int
+        The number of samples to overlap each window.
+    %(verbose)s
+    **csd_kwargs
+        Keyword arguments to pass to :func:`mne.time_frequency.csd_morlet`,
+        :func:`mne.time_frequency.csd_fourier` or
+        :func:`mne.time_frequency.csd_multitaper`.
+
+    Returns
+    -------
+    stc: TFRSourceEstimate
+        The source estimate.
+
+    See Also
+    --------
+    apply_dics
+    apply_dics_epochs
+    apply_dics_csd
+    """
+    from ..time_frequency import csd_morlet, csd_fourier, csd_multitaper
+
+    _check_option('method', method, ('morlet', 'fourier', 'multitaper'))
+    method = dict(morlet=csd_morlet, fourier=csd_fourier,
+                  multitaper=csd_multitaper)[method]
+
+    # ensure data is loaded
+    epochs.load_data()
+
+    if n_csd > epochs.times.size:
+        raise ValueError('`n_csd` cannot be more than the number of '
+                         f'epochs time samples ({epochs.times.size}), '
+                         f'got {n_csd}')
+
+    if n_overlap >= n_csd:
+        raise ValueError('`n_overlap` must be strictly less than `n_csd` '
+                         f'({n_csd}), got {n_overlap}')
+
+    n_orient = 3 if filters['pick_ori'] == 'vector' else 1
+    window_t0s = np.arange(0, epochs.times.size - n_csd, (n_csd - n_overlap))
+    data = np.empty(shape=(window_t0s.size, 1 if average else len(epochs),
+                           filters['n_sources'], n_orient, freqs.size))
```

Comment:
Refresh my mind: is that the dimensions order we agreed on? Wasn't space first?

Reply:
They are reordered below, this is just for convenience/readability